### DISCUSSION TITLE
Poor performance when building simplexml object for configuration

### DISCUSSION THREAD

**Discussion by onlinebizsoft** [OpenMage/magento-lts#2143] [Questions & Answers]
I have a Magento site which has quite many store (around 100 website/store). I'm trying to achieve a performance issue with Magento and I found that PHP (7.x)  has a very poor performance when building the simplexml object from cached XML string
![image](https://user-images.githubusercontent.com/22501570/79102962-f8715f00-7d95-11ea-82bb-7f3fc8dfc790.png)

I was trying to find a way to cache this object directly however simplexml object has no serialize interface, even I tried to use this library but it doesn't support as well
https://github.com/igbinary/igbinary

Has anyone gotten any idea?
@colinmollenhour @Vinai

**Comment by tmotyl** [comment]
Hi
@onlinebizsoft please give more context.
Why do you think this method is a bottleneck in your case? What did you profile and how?
What cache storage do you use?

How many times this method is called?
From where?

I know indeed that simplexml is not very performant but in order to do any optimization the case you have need to be described much better.

**Comment by colinmollenhour** [comment]
I've thought about improvements for SimpleXML many times but I think the only way to improve it would be to replace it (e.g. with native PHP arrays) but that would be a lot of effort and break a lot of code so I don't really see it being feasible.

However, unless I've forgotten something once the config is cached, increasing the number of stores should not really hurt performance since the cache is broken down into different "sections" for each website/store and only the ones used in a request should be loaded, so it should only be loading one store per request.

**Comment by rvelhote** [comment]
I was looking into something recently that's slightly related.
We have a lot of store views in our shop; 98 to be exact :)

Magento does not do `simple_xml_load_string` unless you try to load a configuration for a specific store view but it does [fetch all the configurations from cache](https://github.com/OpenMage/magento-lts/blob/c6f200cb6bc3c4ace0d9b54d0d0df0ad0432ba8f/app/code/core/Mage/Core/Model/App.php#L659). This results, in our case, in 98 calls to Redis.

It's not a huge performance cost for Redis but I can imagine that people that have the cache in the database and lots of storeviews might suffer a bit.

**Comment by colinmollenhour** [comment]
Ahh, nice find! I had never noticed that code before and you are absolutely right, it results in a cache load/save for every store on every request! It only has a handful of keys so they are small payloads but still that is pretty significant for large numbers of stores. It seems that it would be much better to load/save all stores as one cache record since all of them are always loaded anyway.
